031-custom_authentication_in_kafka_brokers.md

```diff
@@ -0,0 +1,199 @@
+          secrets:
+            - name: example
+```
+
+Then, when constructing the broker config, we’ll perform the following tasks:
+
+* `Principal`, if set, is set cluster-wide for all authentication methods. This is a limitation of Kafka, which only allows one principal to be specified for the entire cluster. If set, this override will always be used, even if another authentication type, such as OAuth, purposefully overrides this. 
+* The protocol for this listener would be derived from the `tls` and `sasl`, which then would be appended to `listener.security.protocol.map`. For example, if `tls: true` and `sasl: true`, then the protocol will be `SASL_SSL`. 
+* Each configuration entry under `listener-config` would be pre-appended with `listener.name.<listener-name>`. 
+* `brokerCertChainAndKey` can be specified if a custom certificate wishes to be served from this listener. This automatically creates/shoves the specified certifciate into a keystore automatically for the user, as it's done for other authentication mechanisms today.
+* `tlsTrustedCertificates` functions identically to OAuth’s setting. This is needed for workflows which will use custom mTLS, or mTLS + SASL workflows, as they may require to add additional certificate authorities. Similar to `brokerCertChainAndKey`, a truststore is generated automatically for the user, with their provided certifciate authorities. 
+* `secrets` allows to specify a list of secrets to mount to the pod to the directory `/mnt/strimzi/custom-auth-secrets/<listener-name>`. This is needed for workflows which need additional credentials locally, such as GSSAPI (Kerberos). 
+    
+In addition, this would allow for even more complex setups, where it would be desired to have multiple SASL mechanisms enabled. 
+    
+For example, take the scenario where a user would like to support both scram-sha-512 and their own custom SASL mechanism:
+
+```
+listener.name.<listener-name>.ssl.keystore.location=/tmp/kafka/cluster.keystore.p12
+listener.name.<listener-name>.ssl.keystore.type=PKCS12
+listener.name.<listener-name>.custom-sasl-protocol.sasl.jaas.config=my.custom.loginmodule.CustomLoginModule required;
+listener.name.<listener-name>.scram-sha-512.sasl.jaas.config=org.apache.kafka.common.security.scram.ScramLoginModule required;
+listener.name.<listener-name>.sasl.enabled.mechanisms=CUSTOM_SASL_PROTOCOL,SCRAM-SHA-512
+principal.builder.class=SimplePrincipal.class
+```
+
+All then we need to do simply configure a listener in our Kafka custom resource, with the following properties:
+
+```
+    listeners:
+      - name: custom-auth-listener
+        port: 9093
+        type: internal
+        tls: true
+        configuration:
+          brokerCertChainAndKey:
+            secretName: my-secret
+            certificate: my-certifciate.crt
+            key: my-key.key
+        authentication:
+          type: custom
+          sasl: true
+          principal.builder.class: SimplePrincipal.class
```

Comment:
isn't the principal cluster wide so that user has to set it under the `spec.kafka.config` (as the previous example)?